pandas/tests/base/test_ops.py

```diff
@@ -137,227 +137,238 @@ def setup_method(self, method):
-            if isinstance(o, Index):
-                tm.assert_index_equal(result, Index(values[1:], name="a"))
-            elif is_datetime64tz_dtype(o):
-                # unable to compare NaT / nan
-                tm.assert_extension_array_equal(result[1:], values[2:])
-                assert result[0] is pd.NaT
+            # attach name to klass
+            o = klass(values.repeat(range(1, len(o) + 1)))
+            o.name = "a"
+        else:
+            if isinstance(o, DatetimeIndex):
+                expected_index = orig._values._shallow_copy(values)
             else:
-                tm.assert_numpy_array_equal(result[1:], values[2:])
-
-                assert pd.isna(result[0])
-                assert result.dtype == orig.dtype
+                expected_index = Index(values)
+            expected_index.name = None
+            o = o.repeat(range(1, len(o) + 1))
+            o.name = "a"
+
+        # check values has the same dtype as the original
+        assert o.dtype == orig.dtype
+
+        # check values correctly have NaN
+        nanloc = np.zeros(len(o), dtype=np.bool)
+        nanloc[:3] = True
+        if isinstance(o, Index):
+            tm.assert_numpy_array_equal(pd.isna(o), nanloc)
+        else:
+            exp = Series(nanloc, o.index, name="a")
+            tm.assert_series_equal(pd.isna(o), exp)
+
+        expected_data = list(range(num_values, 2, -1))
+        expected_data_na = expected_data.copy()
+        if expected_data_na:
+            expected_data_na.append(3)
+        expected_s_na = Series(
+            expected_data_na,
+            index=expected_index[num_values - 1 : 0 : -1],
+            dtype="int64",
+            name="a",
+        )
+        expected_s = Series(
+            expected_data,
+            index=expected_index[num_values - 1 : 1 : -1],
+            dtype="int64",
+            name="a",
+        )
 
-            assert o.nunique() == 8
-            assert o.nunique(dropna=False) == 9
+        result_s_na = o.value_counts(dropna=False)
+        tm.assert_series_equal(result_s_na, expected_s_na)
+        assert result_s_na.index.name is None
+        assert result_s_na.name == "a"
+        result_s = o.value_counts()
+        tm.assert_series_equal(o.value_counts(), expected_s)
+        assert result_s.index.name is None
+        assert result_s.name == "a"
+
+        result = o.unique()
+        if isinstance(o, Index):
+            tm.assert_index_equal(result, Index(values[1:], name="a"))
+        elif is_datetime64tz_dtype(o):
+            # unable to compare NaT / nan
+            tm.assert_extension_array_equal(result[1:], values[2:])
+            assert result[0] is pd.NaT
+        elif len(o) > 0:
+            tm.assert_numpy_array_equal(result[1:], values[2:])
+
+            assert pd.isna(result[0])
+            assert result.dtype == orig.dtype
+
+        assert o.nunique() == max(0, num_values - 2)
+        assert o.nunique(dropna=False) == max(0, num_values - 1)
```

Comment:
- comparing against dynamic value depending on the length of the fixture value
- using `max(0, ...)` so tests don't break with the empty index